test-suite/js/src/test/require-sam/org/scalajs/testsuite/jsinterop/CustomJSFunctionTest.scala

```diff
@@ -0,0 +1,112 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js Test Suite        **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2018, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package org.scalajs.testsuite.jsinterop
+
+import scala.scalajs.js
+
+import org.junit.Assert._
+import org.junit.Test
+
+import org.scalajs.testsuite.utils.JSAssert._
+
+class CustomJSFunctionTest {
+
+  import CustomJSFunctionTest._
+
+  @Test def customJSFunctions(): Unit = {
+    val array = js.Array(4, 6, 2, 3).asInstanceOf[js.Dynamic]
+
+    val getAllArgs = new js.Function("...args", "return args;")
+
+    val f1: MyJSFunction1[Int, Int] = { x => x * 2 }
+    assertEquals(42, f1(21))
+    assertJSArrayEquals(js.Array(8, 12, 4, 6), array.map(f1).asInstanceOf[js.Array[Int]])
+
+    assertJSArrayEquals(js.Array(21),
+        getAllArgs.asInstanceOf[MyJSFunction1[Int, js.Array[Int]]](21))
+
+    val f2: MyJSFunctionWithRestParam[String, String] = { args =>
+      args.mkString(", ")
+    }
+    assertEquals("", f2())
+    assertEquals("foo", f2("foo"))
+    assertEquals("foo, bar, baz", f2("foo", "bar", "baz"))
+
+    assertJSArrayEquals(js.Array(5, 4, 8),
+        getAllArgs.asInstanceOf[MyJSFunctionWithRestParam[Int, js.Array[Int]]](5, 4, 8))
+
+    val f3: MyJSFunction1WithRestParam[Int, String, String] = { (x, args) =>
+      "" + x + ": " + args.mkString(", ")
+    }
+    assertEquals("1: ", f3(1))
+    assertEquals("2: foo", f3(2, "foo"))
+    assertEquals("3: foo, bar, baz", f3(3, "foo", "bar", "baz"))
+
+    assertJSArrayEquals(js.Array(5, 4, 8),
+        getAllArgs.asInstanceOf[MyJSFunction1WithRestParam[Int, Int, js.Array[Int]]](5, 4, 8))
+
+    val f4: MyJSFunctionWithDefaultParameter[Int] = { (x, y) => x + y.getOrElse(5) }
+    assertEquals(26, f4(21))
+    assertEquals(31, f4(21, 10))
+
+    assertJSArrayEquals(js.Array(21),
+        getAllArgs.asInstanceOf[MyJSFunctionWithDefaultParameter[js.Array[Int]]](21))
+    assertJSArrayEquals(js.Array(21, 4),
+        getAllArgs.asInstanceOf[MyJSFunctionWithDefaultParameter[js.Array[Int]]](21, 4))
+  }
+
+  @Test def customJSThisFunctions(): Unit = {
+    case class Foo(x: Int)
+
+    val getAllArgs = new js.Function("...args", "return [this].concat(args);")
+
+    val f1: MyJSThisFunction2[Foo, Int, Int, Int] = { (foo, x, y) => foo.x + x + y }
+    assertEquals(30, f1(Foo(5), 21, 4))
+
+    assertJSArrayEquals(js.Array(Foo(5), 21, "hello"),
+        getAllArgs.asInstanceOf[MyJSThisFunction2[Foo, Int, String, js.Array[Any]]](Foo(5), 21, "hello"))
```

Comment:
This is a very indirect way of testing the `this` aspect IMO. Have you considered actually assigning the function to an object and calling it as a method?

Reply:
Hum, I disagree. IMO this is the most direct way to test that we receive exactly the `this` value we want. Assigning the function to an object, then calling it as a method, is less direct than straight calling the function with an explicit `this` value.

Reply:
OK, fair enough. With the additional comment of why we do the `getAllArgs` thing, this is clearer to me.